openai-model-provider/proxy/proxy.go

```diff
@@ -96,6 +108,109 @@ func (s *server) proxyDirector(req *http.Request) {
 	if s.cfg.PathPrefix != "" && !strings.HasPrefix(req.URL.Path, s.cfg.PathPrefix) {
 		req.URL.Path = s.cfg.PathPrefix + req.URL.Path
 	}
+
+	if req.Body == nil || s.cfg.UpstreamHost != openaiBaseHostName || req.URL.Path != chatCompletionsPath {
+		return
+	}
+
+	bodyBytes, err := io.ReadAll(req.Body)
+	if err != nil {
+		fmt.Println("failed to read request body, error: ", err.Error())
+		return
+	}
+
+	var reqBody openai.ChatCompletionRequest
+	if err := json.Unmarshal(bodyBytes, &reqBody); err == nil && isModelO1(reqBody.Model) {
+		if err := modifyRequestBodyForO1(req, &reqBody); err != nil {
+			fmt.Println("failed to modify request body for o1, error: ", err.Error())
+			req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+		}
+	} else {
+		req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+	}
+}
+
+func modifyRequestBodyForO1(req *http.Request, reqBody *openai.ChatCompletionRequest) error {
+	reqBody.Stream = false
+	reqBody.Temperature = nil
+	for i, msg := range reqBody.Messages {
+		if msg.Role == "system" {
+			reqBody.Messages[i].Role = "developer"
+		}
+	}
+	modifiedBodyBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return fmt.Errorf("failed to marshal request body after modification: %w", err)
+	}
+	req.Body = io.NopCloser(bytes.NewBuffer(modifiedBodyBytes))
+	req.ContentLength = int64(len(modifiedBodyBytes))
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Accept-Encoding", "")
+	req.Header.Set("Content-Type", "application/json")
+	return nil
+}
+
+func (s *server) modifyResponse(resp *http.Response) error {
+	if resp.StatusCode != http.StatusOK || resp.Body == nil || resp.Request.URL.Path != chatCompletionsPath || resp.Request.URL.Host != openaiBaseHostName {
+		return nil
+	}
+
+	if resp.Header.Get("Content-Type") == "application/json" {
+		var respBody openai.ChatCompletionResponse
+		rawBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read response body: %w", err)
+		}
+		if err := json.Unmarshal(rawBody, &respBody); err == nil && isModelO1(respBody.Model) {
```

Comment:
So we only have to process the body once.
```suggestion
		if err := json.Decode(resp.Body).Decode(&respBody); err == nil && isModelO1(respBody.Model) {
```